Automattic/vip-go-mu-plugins
# Initial implementation of Files ACL Endpoint

This will allow us to restrict access to files based on different signals from the request (e.g. inherited post status, user permissions, etc.)

See this comment for more details about this functionality: https://github.com/Automattic/vip-go-mu-plugins/pull/1902#issuecomment-755390156

## TODO

- [ ] Finalize endpoint path / name
- [x] Caching
- [x] More tests
- [x] Swap X-Reason with trigger_error

## Checklist

Please make sure the items below have been covered before requesting a review:


## Steps to Test

Since the nginx pieces aren't in place, we can test by calling the PHP script/endpoint directly:

```
curl 'http://vip-go-dev.lndo.site/wp-content/mu-plugins/endpoints/internal-check-file-acl.php' -H 'X-Original-Uri: /testin/wp-content/uploads/path/to/file.png'
```

Here's a script to quickly scaffold some test data using `vip-go-mu-dev` (`vip-dev-env` is missing GD and doesn't work :():

```
WP="lando wp"

IMAGE_ID=`$WP media import --title="Attachment" --porcelain "http://s0.wp.com/i/void.gif"`
IMAGE_URL=`$WP post get $IMAGE_ID --field=guid`

echo "== Attachment URL"
echo $IMAGE_URL
```

You can then pass that into the `X-Original-Uri` header. All requests should return a 202.

#### Invalid requests

If you make a request with invalid data (e.g. missing header, missing wp-content path in header, etc.) it should fail with a 500

If the filter returns an invalid value, it will also fail with a 500.

#### Restricted Files

If you add the following filter it should return a 403 + `X-Private: true`:

```
add_filter( 'vip_files_acl_file_visibility', function( $file_visibility, $sanitized_file_path ) {
        return Automattic\VIP\Files\Acl\FILE_IS_PRIVATE_AND_DENIED;
}, 10, 2 );
```

#### Restricted (But Allowed) Files

If you add the following filter it should return a 202 + `X-Private: true`:

```
add_filter( 'vip_files_acl_file_visibility', function( $file_visibility, $sanitized_file_path ) {
        return Automattic\VIP\Files\Acl\FILE_IS_PRIVATE_AND_ALLOWED;
}, 10, 2 );
```

#### Not Found

```
add_filter( 'vip_files_acl_file_visibility', function( $file_visibility, $sanitized_file_path ) {
        return Automattic\VIP\Files\Acl\FILE_NOT_FOUND;
}, 10, 2 );
```

Should return a 202 as well but the request itself will fail at the nginx level.

#### Subsites

If you create a multisite subsite:

```
lando wp core multisite-convert

lando wp site create --slug=testing
```

You can repeat the earlier scaffolding script with `WP="lando wp --url=--url=http://vip-go-dev.lndo.site/testing/"`

Then, add the `/testing` prefix to `X-Original-Uri`. It should load only the subsite context. You can verify this by adding a subsite check to the filter or `var_dump( get_blog_details() );`
